lib/handlers/error-pages.js

```diff
@@ -1,53 +1,167 @@
   }
 
+  let statusCode = err.status || err.statusCode || 500
+
+  if (statusCode === 401) {
+    setAuthenticateHeader(authMethod, err, locals, res)
+  }
+
+  if (requiresSelectProvider(authMethod, statusCode, req)) {
+    return redirectToSelectProvider(req, res)
+  }
+
   // If noErrorPages is set,
-  // then use built-in express default error handler
+  // then return the response directly
   if (ldp.noErrorPages) {
-    res
-      .status(err.status)
-      .send(err.message + '\n' || '')
-    return
+    sendErrorResponse(statusCode, res, err)
+  } else {
+    sendErrorPage(statusCode, res, err, ldp)
   }
+}
+
+/**
+ * Tests whether a given authentication method requires a Select Provider
+ * page redirect for 401 error responses.
+ *
+ * @param authMethod {string}
+ * @param statusCode {number}
+ * @param req {IncomingRequest}
+ *
+ * @returns {boolean}
+ */
+function requiresSelectProvider (authMethod, statusCode, req) {
+  if (statusCode !== 401) { return false }
+
+  if (!SELECT_PROVIDER_AUTH_METHODS.includes(authMethod)) { return false }
+
+  if (!req.accepts('text/html')) { return false }
+
+  return true
+}
+
+/**
+ * Dispatches the writing of the `WWW-Authenticate` response header (used for
+ * 401 Unauthorized responses).
+ *
+ * @param authMethod {string}
+ * @param err {Error}
+ * @param locals {Object} app.locals hashmap
+ * @param res {ServerResponse}
+ */
+function setAuthenticateHeader (authMethod, err, locals, res) {
+  switch (authMethod) {
+    case 'oidc':
+      Auth.oidc.setAuthenticateHeader(err, locals, res)
+      break
+    case 'tls':
+      Auth.tls.setAuthenticateHeader(locals, res)
+      break
+    default:
+      break
+  }
+}
+
+/**
+ * Sends the HTTP status code and error message in the response.
+ *
+ * @param statusCode {number}
+ * @param res {ServerResponse}
+ * @param err {Error}
+ */
+function sendErrorResponse (statusCode, res, err) {
+  res.status(statusCode)
+  res.send(err.message + '\n')
+}
+
+/**
+ * Sends the HTTP status code and error message as a custom error page.
+ *
+ * @param statusCode {number}
+ * @param res {ServerResponse}
+ * @param err {Error}
+ * @param ldp {LDP}
+ */
+function sendErrorPage (statusCode, res, err, ldp) {
+  let errorPage = ldp.errorPages + statusCode.toString() + '.html'
+
+  return new Promise((resolve) => {
+    fs.readFile(errorPage, 'utf8', (readErr, text) => {
+      if (readErr) {
+        // Fall back on plain error response
+        return resolve(sendErrorResponse(statusCode, res, err))
+      }
 
-  // Check if error page exists
-  var errorPage = ldp.errorPages + err.status.toString() + '.html'
-  fs.readFile(errorPage, 'utf8', function (readErr, text) {
-    if (readErr) {
-      return res
-        .status(err.status)
-        .send(err.message || '')
-    }
-
-    res.status(err.status)
-    res.header('Content-Type', 'text/html')
-    res.send(text)
+      res.status(statusCode)
+      res.header('Content-Type', 'text/html')
```

Comment:
That would be nice, agreed. I've opened issue https://github.com/solid/node-solid-server/issues/508 to implement support for more error formats (other than `text/html`) based on the Accept header.